### DISCUSSION TITLE
Event modulation Index and Modulation significance level

### DISCUSSION THREAD

**Discussion by jesusmartin-cortecero** [petersenpeter/CellExplorer#82] [Q&A]
Hi Peter.

I am trying to get some information from  PSTHs and rasters using Cell Explorer and i have seen that Cell Explorer gives you the "modulation Index". How is this parameter calculated? 
I have seen in several papers that the values goes from -1 to 1, however Cell Explorer gives me values up to 200. 

I would like also to know how Cell Explorer calculates the "Modulation Significance level",  I could not find this information in the Website. 

Thank you very much in advance

Jesus

**Comment by petersenpeter** [comment]
Hi Jesus

Take a look at the `calc_PSTH.m` function that does this. It is the ratio between the average response divided by the average baseline pre-stimulation: https://github.com/petersenpeter/CellExplorer/blob/69c4a1eca9aba5ec8dc6eaee6f75fdc96c7d283e/calc_CellMetrics/calc_PSTH.m#L101
```m
modulationIndex = mean(PSTH_out(binsEvents,:))./mean(PSTH_out(binsPre,:));
```

The modulation significance level is a KS-test also calculated in the same script:
```m
modulationSignificanceLevel = [];
for i = 1:size(PSTH_out,2)
    [~,p_kstest2] = kstest2(PSTH_out(binsEvents,i),PSTH_out(binsPre,i));
    modulationSignificanceLevel(i) = p_kstest2;
end
```

**Reply by petersenpeter** [reply]
Could you share a link to the modulation index you refer to? If it can be generalized, I can implement it into the pipeline. 

/ Peter

**Reply by jesusmartin-cortecero** [reply]
Thank you very much for the information. 

Regarding the modulation Index, although  it varies a bit among different labs or projects, the one i meant can be found in these papers
.

https://www.nature.com/articles/s41467-019-10268-z
https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4764557/

i wonder which modulation index is easier to interpret, what do you think? 

Jesus

**Reply by jesusmartin-cortecero** [reply]
i am trying to understand how CellExplorer does to calculate de modulation Index and significance level and i am having problems to understand "calc_PSTH"  

I would like to specify the window to take into consideration for the analysis of the Modulation Index or significance level. For example, although my stimulus last 0.05 seconds, i want to use the first 0.015 sec  after onset to determine is an early response compatible with a peripheral input. How can i specify the window to use for Modulation Index or significance level? 
Should i set the inputparser "duration" in 0.015?

thank you very much in advance

Jesus

**Comment by petersenpeter** [comment]
Hi Jesus

I revisited this and made some modifications and improvements. The **modulation index** is now scaled between -1 and 1 as in your second reference, defined as the difference between the stimulation interval and the pre-stimulation interval, divided by their sum). I added a second measure called the **stimulation ratio** defined as the stimulation interval divided by the pre-stimulation interval. These are calculated from the averages of the intervals. 

I updated the documentation to reflect this: https://cellexplorer.org/datastructure/standard-cell-metrics/#event-metrics

I also updated the preferences file of the Processing Module to allow you to define custom parameters for the PSTH: https://github.com/petersenpeter/CellExplorer/blob/master/preferences_ProcessCellMetrics.m

```m
% PSTH for ripples (your custom preferences for your own events, you can rename the struct name to your events name)
preferences.psth.ripples.binCount = 100;           % how many bins (for half the window)
preferences.psth.ripples.alignment = 'onset';    % alignment of time ['onset','center','peaks','offset']
preferences.psth.ripples.binDistribution = [0.33,0.34,0.33];   % How the bins should be distributed around the events, pre, during, post. Must sum to 1
preferences.psth.ripples.duration = 0.025;   % duration of PSTH (for half the window - used in CCG) [in seconds]
preferences.psth.ripples.smoothing = 5;       % any gaussian smoothing to apply? units of bins.
preferences.psth.ripples.percentile = 99;      % if events do not have the same length, the event duration can be determined from the percentile of the distribution of events
```

In your case, you could set the duration to 0.025 and the binDistribution to [0.34,0.33,0.33] this will give you a pre-stimulation baseline of 17ms and a stimulation interval of 16.5ms.

I am working on a better method that allows you to set specific intervals. I will let you know when it is ready.

**Reply by jesusmartin-cortecero** [reply]
Hi Peter 

Many thanks for the update. now i can see the modulation index is from -1 to 1.

There is something i  dont understand yet, if the duration is 25ms why i have 17ms pre and 16.5ms stimulation interval? in total this is 31.5ms. right? if i set distribution to [0.33 0.34 0.33] i would expect it will give me 8.25ms  pre and 8.5ms stimulation interval and 8.25ms would be post stimulation.  

i
![image](https://user-images.githubusercontent.com/53047456/201334557-a1b24a6b-e900-4f30-82a9-84b2d4147a48.png)
here is an example of psth setting the "duration" and "maxwindow" to 25ms. however the displayed window is around 70ms.

**Reply by petersenpeter** [reply]
The duration parameter is really half the window duration, so in your case, the full window is 50ms. It is named as it is because it uses the CCG function (which operates with the half window size) for calculating the PSTHs. It is in the description of the parameters, but I can see it is a misleading name.

**Reply by jesusmartin-cortecero** [reply]
Hi peter,

How can i select the clusters to be included in the "calc_PSTH" function? 

Thanks in advance

**Reply by petersenpeter** [reply]
It is calculated for all units. If you want to select, you would need to create a spikes struct with a subset of the units.

**Comment by jesusmartin-cortecero** [comment]
Hi peter when i get the Psth, using the script calc_Psth, i am trying to understand how the responsive curve is calculated, i see very high values. so i guess is in Hz? 

Thanks in advance
